### PR TITLE
Fixed Bloch-Messiah bug arising when singular values were degenerate

### LINKED ISSUE #14: Bloch-messiah decomposition sometimes returns decomposed matrices with permuted rows and columns
The Bloch-Messiah decomposition should return two orthogonal symplectic matrices O1 and O2. In some cases, for instance when the squeezing values are degenerate, the `bloch_messiah` function returns the correct orthogonal symplectic matrices, up to a permutation of rows and columns.

Minimal non-working example:

```python
>>> S = np.array([[ 1.,  0.,  0.,  0.],
                  [ 1.,  1.,  0.,  0.],
                  [ 0.,  0.,  1., -1.],
                  [ 0.,  0.,  0.,  1.]])

>>> O1, Z, O2 = bloch_messiah(S)
```

* *Expected behavior:* 

```python
>>> O1 @ sympmat(2) @ O1.T == sympmat(2)
True
```

* *Actual behavior:*

```python
>>> O1 @ sympmat(2) @ O1.T == sympmat(2)
False
>>> O1 @ sympmat(2) @ O1.T
array([[-0., -1.,  0.,  0.],
       [ 1., -0.,  0.,  0.],
       [ 0.,  0.,  0., -1.],
       [ 0.,  0.,  1., -0.]])
```

A possible solution would be to check whether the returned matrices from `bloch_messiah` are symplectic, and if not, calculate the required row and column permutations.

### PR DESCRIPTION
**Pull request by ishdhand**
**Description of the Change:**
Replaced the permutation matrix originally applied on output to permute it into canonical symplectic form with a rotation matrix that is constructed using SVD. The earlier approach did not return matrices in the canonical symplectic form if one or more of the Bloch-Messiah singular values were degenerate; this is fixed now.

**Benefits:**
Bloch-Messiah now works for input matrices with degenerate singular values.

**Related GitHub Issues:**
Fixes #14, which was an instance of degenerate singular values (both equal to the φ = 1.61803)

### CODE REVIEW

**Comment by ishdhand** [comment]
I just found some stray input parameters that give unexpected results. Next steps for me are to write tests to better identify this issue and fix it.

**Comment by josh146** [comment]
Thanks @ishdhand! It looks great. Let me know when you have finished working on the edge case test, and I will start the code-review process.

**Comment by ishdhand** [comment]
Thanks! The latest commit 54709d2 fixes that edge case so good to go from my end (after the codacy thumbsup).

**Comment by co9olguy** [comment]
Awesome work @ishdhand

File: strawberryfields/decompositions.py [hunk 1]
```diff
@@ -278,16 +278,29 @@ def bloch_messiah(S, tol=10):
...
     ## (and their corresonding inverses) to get the canonical symplectic form
     qomega = np.transpose(ut) @ (omega) @ ut
-    perm1 = list(range(2*n))
+    st = pmat @ np.diag(ss) @ pmat
 
-    for i in range(n):
-        if qomega[i, i+n] < 0:
-            perm1[i] = i+n
-            perm1[i+n] = i
+    # Identifying degenrate subspaces
```
Comment (co9olguy) [inline strawberryfields/decompositions.py]:
```suggestion
    # Identifying degenerate subspaces
```

File: strawberryfields/decompositions.py [hunk 2]
```diff
@@ -278,16 +278,29 @@ def bloch_messiah(S, tol=10):
...
+    result = []
+    for _k, g in groupby(np.diag(st)[:n]):
+        result.append(list(g))
+
+    stop_is = list(np.cumsum([len(res) for res in result]))
+    start_is = [0] + stop_is[:-1]
+
+    # Rotation matrices (not permutations) based on svd.
+    # See Appending B2 of Serafini's book for more details.
```
Comment (co9olguy) [inline strawberryfields/decompositions.py]:
```suggestion
    # See Appendix B2 of Serafini's book for more details.
```

**Comment by co9olguy** [comment]
@ishdhand you mentioned adding new tests. Is this still part of the plan?

**Comment by ishdhand** [comment]
@co9olguy, thanks for nudging me to write tests, which helped identify more edge cases. a91122a includes the tests and the resolved edge cases. e9e630a fixes the typos that you caught. Thanks!

**Comment by co9olguy** [comment]
Nice @ishdhand! 
@josh146 @nquesada can you guys provide a more technical code review on this PR?

**Comment by ishdhand** [comment]
Thanks for going through the code, you guys! :)
